[AutoDiff upstream] Add the _Differentiation module.

dan-zheng opened the pull request:
The `_Differentiation` module is the experimental support library for
differentiable programming. It is built when the build-script flag
`--enable-experimental-differentiable-programming` is enabled.

The `Differentiable` protocol generalizes all types that work with
differentiation. It is a core piece of the differentiable programming
project. Other parts depending on the `Differentiable` protocol will
be upstreamed piece by piece.

The `Differentiable` protocol is compiler-known and will be used during
type-checking, SILGen, and the SIL differentiation transform.

---

[Forum discussion on upstreaming differentiable programming.](https://forums.swift.org/t/upstreaming-differentiable-attribute-and-differentiable-protocol/26821)

rxwei commented on stdlib/public/core/Differentiable.swift:
```diff
@@ -0,0 +1,43 @@
+}
+
+public extension _Differentiable where TangentVector == Self {
+  mutating func move(along direction: TangentVector) {
```
```suggestion
  @_alwaysEmitIntoClient
  mutating func move(along direction: TangentVector) {
```

dan-zheng commented:
Hi @DougGregor: could you please help review this patch (or forward to appropriate reviewers)?

Compilation of the `_Differentiable` protocol is gated by the `--enable-experimental-differentiable-programming` `build-script` flag, which is on by default to ensure testing. The design and naming will go through Swift Evolution.

rxwei commented:
This patch is a direct dependency of many upcoming patches that can be done in parallel, including [TF-828](https://bugs.swift.org/browse/TF-828), [TF-829](https://bugs.swift.org/browse/TF-829), and [TF-830](https://bugs.swift.org/browse/TF-830). Any review on this patch will be greatly appreciated!

DougGregor commented:
Why is this going into the standard library vs. an auto-diff support library? Is it truly the only library piece for auto-diff?

rxwei commented:
@DougGregor It is because we don't yet support retroactive derivative registration, and we will need to define derivatives for `Float` for testing purposes. Adding a derivative requires the use of `@differentiating` attribute, whose type checking requires the `Differentiable` protocol.

That said, we now aim to finish support for retroactive derivative registration this week with [SIL differentiability witnesses](https://github.com/apple/swift/pull/27487). So we can check everything into the `_Differentiation` module and not make modifications to the `Swift` module today.

rxwei commented:
@dan-zheng Could you update the PR to add the `Differentiable` protocol (with no underscores) to a `_Differentiation` module in stdlib/public?

@DougGregor We'll ping you when it's ready for review again in an hour or so.

dan-zheng committed:
[AutoDiff upstream] Add the _Differentiation module.

The `_Differentiation` module is the experimental support library for
differentiable programming. It is built when the build-script flag
`--enable-experimental-differentiable-programming` is enabled.

The `Differentiable` protocol generalizes all types that work with
differentiation. It is a core piece of the differentiable programming
project. Other parts depending on the `Differentiable` protocol will
be upstreamed piece by piece.

The `Differentiable` protocol is compiler-known and will be used during
type-checking, SILGen, and the SIL differentiation transform.

dan-zheng commented:
> @dan-zheng Could you update the PR to add the `Differentiable` protocol (with no underscores) to a `_Differentiation` module in stdlib/public?

Done in https://github.com/apple/swift/pull/27511/commits/5dc37c30a828a3685c65dd500d8f8c889ec8978d.

dan-zheng commented:
Hi @DougGregor (and others): could you please help re-review this patch?

It adds a flag-gated experimental `_Differentiation` module to `stdlib/public` - all contents are experimental and subject to change. I hope the patch isn't controversial - it's blocking our other autodiff upstreaming patches, so I hope it's okay if we merge this within 24 hours if no other reviewers respond.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@DougGregor: ping in case you have time for review.

DougGregor commented on include/swift/AST/KnownIdentifiers.def:
```diff
@@ -53,6 +53,7 @@ IDENTIFIER(decode)
 IDENTIFIER(decodeIfPresent)
 IDENTIFIER(Decoder)
 IDENTIFIER(decoder)
+IDENTIFIER_WITH_NAME(Differentiation, "_Differentiation")
```
Note that you can use `IDENTIFIER_` for this case (it's not a big deal).

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27511/commits/5aececee414f9a56adfed3f751a24df2d4a08c5e.

Locally verified that `ASTContext::getProtocol(KnownProtocolKind::Differentiable)` works (returns non-null pointer).

DougGregor commented on test/Differentiation/differentiable_protocol.swift:
```diff
@@ -0,0 +1,33 @@
+// RUN: %target-typecheck-verify-swift
+
```
Hmm. I think you'll need a `lit` configuration option `autodiff` and ` // REQUIRES: autodiff` on this test?

dan-zheng replied:
Thanks for the catch! Added a `lit` option called `differentiable_programming` in https://github.com/apple/swift/pull/27511/commits/5aececee414f9a56adfed3f751a24df2d4a08c5e.

DougGregor reviewed (approved):
LGTM

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
There appears to be an orthogonal SourceKit-LSP test failure:
```
10:56:44 $ "/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/llvm-linux-x86_64/bin/FileCheck" "--input-file" "/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/test-sourcekit-lsp/Output/test-sourcekit-lsp.py.tmp.run-log" "/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift-integration-tests/test-sourcekit-lsp/test-sourcekit-lsp.py"
10:56:44 # command stderr:
10:56:44 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift-integration-tests/test-sourcekit-lsp/test-sourcekit-lsp.py:124:11: error: CHECK: expected string not found in input
10:56:44  # CHECK: "items":[
10:56:44           ^
10:56:44 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/test-sourcekit-lsp/Output/test-sourcekit-lsp.py.tmp.run-log:55:253: note: scanning from here
10:56:44 {"jsonrpc":"2.0","id":4,"result":{"items":[{"kind":2,"label":"foo()","insertTextFormat":1,"detail":"Void","insertText":"foo()","filterText":"foo()"},{"kind":14,"label":"self","insertTextFormat":1,"detail":"Lib","filterText":"self","insertText":"self"}],"isIncomplete":false}}Content-Length: 337
10:56:44                                                                                                                                                                                                                                                             ^
10:56:44 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/test-sourcekit-lsp/Output/test-sourcekit-lsp.py.tmp.run-log:57:78: note: possible intended match here
10:56:44 {"method":"textDocument\/publishDiagnostics","jsonrpc":"2.0","params":{"diagnostics":[],"uri":"file:\/\/\/home\/buildnode\/jenkins\/workspace\/swift-PR-Linux-smoke-test\/branch-master\/buildbot_linux\/none-swift_package_sandbox_linux-x86_64\/tests\/test-sourcekit-lsp\/Output\/test-sourcekit-lsp.py.tmp.dir\/pkg\/Sources\/clib\/clib.c"}}
10:56:44                                                                              ^
10:56:44 
10:56:44 error: command failed with exit status: 1
10:56:44 
10:56:44 --
10:56:44 
10:56:44 ********************
10:56:48 Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
10:56:52 Testing Time: 12.59s
10:56:52 ********************
10:56:52 Failing Tests (1):
10:56:52     swift-package-tests :: test-sourcekit-lsp/test-sourcekit-lsp.py
```

I'll rebase on top of master and rerun tests.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
Note since last review: the `test/Differentiation` subdirectory is renamed to `test/AutoDiff` for consistency with [`test/AutoDiff`](https://github.com/apple/swift/tree/tensorflow/test/AutoDiff) on `tensorflow` branch.

dan-zheng commented:
@swift-ci Please smoke test
